### ISSUE TITLE
Error in slot(value, what) : no slot of name "reducedDims"

### ISSUE THREAD

**Issue by mimi3421** [RGLab/MAST#119]
When package _SingleCellExperiment_ is updated to 1.7.1, the following error will raise when _FromMatrix_ is called in MAST 1.11.3 because "reducedDims" is removed in new version of _SingleCellExperiment_: 

> "Error in slot(value, what) : no slot of name "reducedDims" for this object of class "SingleCellExperiment" 

I find a temporal hack to this problem which is re-defining _SingleCellAssay_ before calling _FromMatrix_:

```
Mandatory_Featurevars <- character()
Mandatory_Cellvars <- character()
setClass('SingleCellAssay', contains='SingleCellExperiment',package='MAST',
         slots=list(cmap='character', fmap='character',reducedDims='character'),
         prototype=list(cmap=Mandatory_Cellvars,
                        fmap=Mandatory_Featurevars))
```

**Comment by amcdavid** [comment]
This looks like a mismatch of bioconductor versions. What does
`BiocManager::valid()` report?

**Comment by mimi3421** [comment]
My bioconductor versions is Bioconductor version '3.9'. I also think that the same error has been reproduced by *ryandhindsa* in https://github.com/RGLab/MAST/issues/115 . 

Everything will be OK if I force my _R3.6_ to load 1.4.1 version of _SingleCellExperiment_ in the _R3.5_ library folder. But it will affect other packages that depend on the newer version of _SingleCellExperiment_ or _S4Vectors_. 

Thanks for your great job on MAST. Anyway I think that _R_ doesn't load dependent packages in private namespace is so foolish especially when some compatibility issues happen considering to the version change.

**Comment by amcdavid** [comment]
You never said what `BiocManager::valid()` says, but I'm guessing you've got an unholy mixture of versions.   Maybe what you did makes the error go away, but no promises it's not introducing other problems that aren't being signalled with an error, since you are mixing SingleCellExperiment from Bioconductor 3.8 and if you installed MAST from github, MAST from the development version of Bioconductor, which will be released in a few weeks as 3.10.  I'd bet you $50 if you use `BiocManager::install()` to install everything and get it so that `BiocManager::valid()` reports true, the errors will go away. 

However, you may be reluctant to do that, because you want to pin the version of packages to keep other analyses reproducible.   This does get tricky in R. The best solution I have found is to explicitly manage your package libraries on a per-project basis.  Some advice suitable for MacOS can be found [here](https://amcdavid.github.io/2017/09/hygiene-of-the-R-library-path) and [here](https://amcdavid.github.io/2016/09/joys-of-multiple-r-environments).  I maintain an `.Renviron` file in every project directory which points to a specific folder that contains a given version of Bioconductor.  Once that version has been superseded, I don't do any upgrading of existing packages in it.

**Comment by amcdavid** [comment]
Closed in c8999fe

**Comment by mimi3421** [comment]
Thanks for your suggestions. I will wait for the stable release in Bioconductor 3.10.
